### DISCUSSION TITLE
How to start new project with atmega-hal with bare MCU (no arduino)

### DISCUSSION THREAD

**Discussion by T0maas** [Rahix/avr-hal#615] [General]
How to start new project and what do I need to write to Cargo.toml? I've just downloaded cargo on my linux distro so I have fresh instalation of rust.
I want to use atmega-hal (not arduino-hal, because I need better control over MCU: timers, prescalers, PWM, ADC, ...), and I have those questions:

- Where to specify target MCU (for example atmega328p or attiny85 or atmega48 or atmega2560) ?
- Where to specify F_CPU (I want to use 8MHz internal oscilator)?
- I did not find atmega-hal or avr-hal in `cargo search` how to add them to Cargo.toml?
- Is running cargo build enough to produce avr compatible binary or do I need to specify some additional arguments?
- Do I need to manually download some rust toolchains?
- If I have avrdude (and I am used to work with it) do I need ravedude or is it optional?
- Is there some **getting started** documentation which describes above questions?

I've read some examples, but they are using relative paths like this:
```
[dependencies.atmega-hal]
path = "../../mcu/atmega-hal/"
features = ["atmega2560"]
```
And I cant use this in my project (where Cargo.toml is in project root directory). Do I need to download this repo somewhere and point path to this directory inside my project?

I have this setup:
```
Default host: x86_64-unknown-linux-gnu
rustup home:  /home/tom/.rustup

stable-x86_64-unknown-linux-gnu (default)
rustc 1.83.0 (90b35a623 2024-11-26)
```

I've found somewhere that I need to run ` rustup target add avr-atmega328p` but it ended with error:
```
error: toolchain 'stable-x86_64-unknown-linux-gnu' does not support target 'avr-atmega328p'
```

I also tried this:
```cargo doc --features atmega328p --open```

but it wrote:
```
error: the package 'atmega328example' does not contain this feature: atmega328p
```
where atmega328example is name of my project directory. It's fresh directory just created using `cargo init`.

I'tried `cargo generate --git https://github.com/Rahix/avr-hal-template.git` but it has no option for bare atmega328p (only arduino UNO was closest)  and it uses arduino-hal instead of atmega-hal.

**Comment by stappersg** [comment]
On Sun, Dec 29, 2024 at 06:19:07AM -0800, T0maas wrote:
> 
> I'tried
>   `cargo generate --git https://github.com/Rahix/avr-hal-template.git`
> but it has no option for bare atmega328p (only arduino UNO was closest)
> and it uses arduino-hal instead of atmega-hal.
> 


```
cargo generate --git https://github.com/Rahix/avr-hal-template.git
cd projectname
cargo build
```

From that "known good" start making changes.
Use `cargo build` to verify it still build.
And at some point deploy to the actual hardware.


Groeten
Geert Stappers
-- 
Silence is hard to parse

**Reply by T0maas** [reply]
> From that "known good" start making changes.

But how? Which changes? What do I need to write to Cargo.toml to make correct configuration for baremetal atmega328p with 8MHz clock?

**Reply by Rahix** [reply]
The clock rate is defined like this, for example: https://github.com/Rahix/avr-hal/blob/4bf713c7361b97d26eaf10b7a683a6510ab0dd80/examples/atmega2560/src/bin/atmega2560-i2cdetect.rs#L7-L8

**Comment by stappersg** [comment]
On Mon, Dec 30, 2024 at 12:16:46PM -0800, T0maas wrote:
> > From that "known good" start making changes.
> 
> But how?

Yes, so true.


> Which changes?

Your changes to a known good,
your changes that bring you closer to your goal.


> What do I need to write to Cargo.toml to make correct configuration
> for baremetal atmega328p with 8MHz clock?

https://github.com/Rahix/avr-hal/discussions?discussions_q=clock
did yield https://github.com/Rahix/avr-hal/discussions/318 which has
title "Target microcontroller directly instead of an arduino board"

Please advice how to improve our documentation. (Karma bonus points
for merge requests.)

And please accept my apology.  I'm sorry that I did not ignore the
original posting. http://www.catb.org/~esr/faqs/smart-questions.html
says I should have ignored it.


Groeten
Geert Stappers
-- 
Silence is hard to parse

**Comment by T0maas** [comment]
Is there some complex documentation? I found only this: https://rahix.github.io/avr-hal/atmega_hal/index.html. I tried to run `cargo doc --features atmega328p --open ` inside avr-hal-template and i've got error:
```
error: none of the selected packages contains these features: atmega328p
```

**Reply by Rahix** [reply]
You need to run this command from the `atmega-hal` or `attiny-hal` directories.

**Comment by Rahix** [comment]
Unfortunately the bare-metal workflow isn't documented well at the moment.

Check the [Quickstart Guide](https://github.com/Rahix/avr-hal#quickstart) and especially the [avr-hal-template](https://github.com/Rahix/avr-hal-template).  I suggest setting up for an Arduino board similar to your target MCU.

Then check the "bare AVR" examples crate here: https://github.com/Rahix/avr-hal/tree/main/examples/atmega2560

You will have to adapt the template you just generated from avr-hal-template according to the stuff you find in that example directory.  Most importantly, switch from depending on `arduino-hal` to depending on `atmega-hal` (or `attiny-hal`), using the same git dependency.
